src/lib/adapters/WebDav.js

```diff
@@ -0,0 +1,400 @@
+
+        if (response.status !== 200) {
+            return {
+                'status' : response.status,
+                'db' : new Map ()
+            };
+        }
+
+        let bookmark_array = await response.json();
+        let server_db = new Map ();
+
+        bookmark_array.forEach ( (bm) => {
+            server_db.set(bm.id, {
+                id: bm.id
+                , url: bm.url
+                , title: bm.title
+                , path: bm.path
+            });
+        });
+
+console.log ("response out");
+console.log (server_db);
+
+        return {
+            'status' : response.status,
+            'db' : server_db
+        };
+    }
+
+    async syncStart () {
+        console.log ("syncStart: started");
+        try {
+            let resp = await this.pullFromServer ();
+            this.db = resp.db;
+
+console.log ("resp");
+console.log (resp);
+
+            if (resp.status !== 200)
+            {
+                if (resp.status !== 404)
+                {
+                    throw new Error('Failed to fetch bookmarks :' + resp.status + ":");
+                }
+            }
+        } catch (e) {
+            console.log ("caught error");
+            console.log (e);
+
+            this.db = new Map ();
+        }
+
+        console.log ("syncStart: completed");
+    }
+
+    async pullBookmarks () {
+        console.log('Fetching bookmarks', this.server)
+
+        let server_db;
+        try {
+            let resp = await this.pullFromServer ();
+            server_db = resp.db;
+
+            if (resp.status !== 200)
+            {
+                if (resp.status !== 404)
+                {
+                    throw new Error('Failed to fetch bookmarks :' + resp.status + ":");
+                }
+            }
+        } catch (e) {
+            console.log ("caught error");
+            console.log (e);
+
+            server_db = new Map ();
+        }
+
+        let myBookmarks = Array.from(server_db.values())
+            .map(bm => {
+                    return new Bookmark(bm.id, null, bm.url, bm.title, bm.path)
+                    })
+
+        console.log('Received bookmarks from server')
+        console.log (myBookmarks);
+
+        return myBookmarks
+    }
+
+    async getBookmark (id, autoupdate) {
+        console.log('Fetching single bookmark', this.server);
+        let bm = this.db.get(id);
+        if (!bm) {
+            throw new Error('Failed to fetch bookmark');
+        }
+        let bookmark = new Bookmark(bm.id, null, bm.url, bm.title, bm.path);
+        return bookmark;
+    }
+
+    async createBookmark (bm) {
+        console.log('Create single bookmark', bm, this.server)
+            if (!~['https:', 'http:', 'ftp:'].indexOf(url.parse(bm.url).protocol)) {
+                return false
+            }
+
+        let highestID = 0;
+        this.db.forEach ( (value, key) => {
+            if (value && value.id && highestID < value.id)
+            highestID = value.id;
```

Comment:
a sort by id would probably be faster, but that's being picky :P

Reply:
Also, since the user can also delete bookmarks, it might be best to have the counter persisted in the file, otherwise, if the last entry is deleted and at the same time a new bookmark is created, it gets the same id, causing other clients to think it was changed, rather then two independent operations. Not sure if that's actually a problem, but let's not make it one :)

Reply:
I will look at improving this code, now that I got it working in general.  I am new to this version of Javascript and the new "functional" style programming is new to me.